src/Common/src/Internal/Cryptography/HashProviderCng.cs

```diff
@@ -0,0 +1,145 @@
+        //
+        //   - "key" activates MAC hashing if present. If null, this HashProvider performs a regular old hash.
+        //
+        public HashProviderCng(string hashAlgId, byte[] key)
+        {
+            BCryptOpenAlgorithmProviderFlags dwFlags = BCryptOpenAlgorithmProviderFlags.None;
+            if (key != null)
+            {
+                _key = key.CloneByteArray();
+                dwFlags |= BCryptOpenAlgorithmProviderFlags.BCRYPT_ALG_HANDLE_HMAC_FLAG;
+            }
+
+            _hAlgorithm = Interop.BCrypt.BCryptAlgorithmCache.GetCachedBCryptAlgorithmHandle(hashAlgId, dwFlags);
+
+            SafeBCryptHashHandle hHash = null;
+            NTSTATUS ntStatus = Interop.BCrypt.BCryptCreateHash(_hAlgorithm, out hHash, IntPtr.Zero, 0, key, key == null ? 0 : key.Length, BCryptCreateHashFlags.BCRYPT_HASH_REUSABLE_FLAG);
+            if (ntStatus == NTSTATUS.STATUS_INVALID_PARAMETER)
+            {
+                // If we got here, we're running on a downlevel OS (pre-Win8) that doesn't support reusable CNG hash objects. Fall back to creating a 
+                // new HASH object each time.
+                ResetHashObject();
+            }
+            else if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+            {
+                throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+            }
+            else
+            {
+                _hHash = hHash;
+                _reusable = true;
+            }
+
+            unsafe
+            {
+                int cbSizeOfHashSize;
+                int hashSize;
+                ntStatus = Interop.BCrypt.BCryptGetProperty(hHash, Interop.BCrypt.BCryptPropertyStrings.BCRYPT_HASH_LENGTH, &hashSize, sizeof(int), out cbSizeOfHashSize, 0);
+                if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                    throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+                _hashSize = hashSize;
+            }
+            return;
+        }
+
+        public sealed override void AppendHashDataCore(byte[] data, int offset, int count)
+        {
+            unsafe
+            {
+                fixed (byte* pRgb = data)
+                {
+                    NTSTATUS ntStatus = Interop.BCrypt.BCryptHashData(_hHash, pRgb + offset, count, 0);
+                    if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                        throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+                }
+            }
+        }
+
+        public sealed override byte[] FinalizeHashAndReset()
+        {
+            byte[] hash = new byte[_hashSize];
+            NTSTATUS ntStatus = Interop.BCrypt.BCryptFinishHash(_hHash, hash, hash.Length, 0);
+            if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+
+            ResetHashObject();
+            return hash;
+        }
+
+        public sealed override void Dispose(bool disposing)
+        {
+            if (disposing)
```

Comment:
I know you have a comment to this effectin GetCachedBCryptAlgorihmHandle, but I think it'd be worth adding a comment here that we explicitly don't dispose of _hAlgorithm because it's a cached object used by other instances.